experimental/auth-logic/top_level.go

```diff
@@ -36,7 +36,7 @@ func verifyRelease(appName, endorsementFilePath, provenanceFilePath string) (str
 	}
 	endorsementStatement, err := wrappers.EmitStatementAs(
 		wrappers.Principal{
-			Contents: fmt.Sprintf(`"%s::EndorsementFile"`, endorsementAppName),
+			Contents: fmt.Sprintf(`"%s::EndorsementFile"`, wrappers.SanitizeName(endorsementAppName)),
```

Comment:
FYI if we used more type safe objects to represent principals and statements (as per https://github.com/project-oak/transparent-release/pull/46#discussion_r845316236), we may only have to do this escaping in a single place (i.e. when converting the principal to string). Currently there is nothing that enforces that names are sanitized correctly across the codebase, it needs to be done for each occurrence.

Reply:
I don't think type safety has any effect on whether names are sanitized everywhere or not unless the interface for authorization logic (which is meant to generalize to more than one project) is tied to this one specific project (because what sanitization needs to be done is specific to this project).

Because sanitization is specific to this project, we also need a public constructor for principals that takes a string (`ConstructPrincipal: String -> Principal`) and doesn't sanitize.

Your point is to add `CompoundPrincipal: Principal -> Principal -> Principal`, and your claim is that this ensures sanitization happens everywhere.

Sanitize can either be `Sanitize: String -> String` or it can be `Sanitize: String -> Principal`. Presumably your proposal works by assuming it returns a principal (and that a compound principal is passed to `EmitStatement`).

EmitStatement is `(Principal, Unattributed) -> Statement`. I think your proposal is meant to work by forcing a compound statement to be used in this position, but it doesn't. Even with these functions available, the programmer can still write either:

`EmitStatement(CompoundPrincipal(Sanitize(<AppName>), ConstructPrincipal("::Endorsement")), statement)`

or

`EmitStatement(CompoundPrincipal(ConstructPrincipal(<AppName>), ConstructPrincipal("::Endorsement")), statement)`

or

`EmitStatement(ConstructPrincipal(<AppName> + "::Endorsement"), statement)`

And 2 out of 3 of those don't do sanitization. So your idea does not achieve its stated purpose, and despite not achieving its stated purpose, it does make the code more complicated.

If the goal really is to ensure that sanitization happens everywhere (and in doing so tying the interface to this specific project), a solution that would actually work is to just move the sanitization logic into the implementation of the `String -> Principal` constructor.

Reply:
Hmmm... I guess the sanitization that needs to happen is not all that specific to transparent release. So perhaps it is best to just move the sanitization logic into the constructor of principal and then it really will happen everywhere. WDYT?

Reply:
> Hmmm... I guess the sanitization that needs to happen is not all that specific to transparent release. So perhaps it is best to just move the sanitization logic into the constructor of principal and then it really will happen everywhere. WDYT?

yes that's what I meant; have a type safe constructor for Principal, and then also type constructors for compound principals and statements etc. I think it should be very rare that we pass actual strings around.

> `EmitStatement(CompoundPrincipal(ConstructPrincipal(<AppName>), ConstructPrincipal("::Endorsement")), statement)`
>
> or
>
> `EmitStatement(ConstructPrincipal(<AppName> + "::Endorsement"), statement)`

These should be different principals (if at all); currently, as you noted, the sanitize function is not even injective, so we probably need to somehow escape special characters instead of purely removing them (or we could forbid some entirely, e.g. whitespaces, or `::`, and prevent creating a principal with those in the first place).

Reply:
I think the idea is to have more custom-defined Go structs so that instead of changing 7-8 files for a behaviour like this, you'd modify only a single file.

Reply:
@tiziano88 nothing about your reply shows how type safety has any affect on whether sanitization takes place. So I am still pretty convinced it does not have this affect.

What you are both describing is more like an E-DSL, which is like a whole separate project that we have talked about in a few other cases. It might have other merits, but I am pretty sure it does not fix the problem you pointed out in this comment :) Do you just want me to make some issue for a go `E-DSL` ?

What about just moving this logic into the constructor for the principal? Nobody commented on that :)

Reply:
What is E-DSL?

Reply:
Embedded domain specific language. It's a way of implementing something that resembles a programming language by instead implementing a library in some other language, sometimes called the host language. (The E-DSL is "embedded" in the host language). Examples of E-DSLs include tensorflow, [Chisel](https://en.wikipedia.org/wiki/Chisel_(programming_language)), and [Halide](https://en.wikipedia.org/wiki/Chisel_(programming_language)).

Reply:
I don't think that is what we are talking about here. My understanding is that we just need some refactoring. 

For instance, you already have a `ValidatedEndorsement` struct with a `Name` field, which is a string. Instead of a string it could be custom-defined `SanitizedName` type that takes a string and stores a sanitized version of it. You'd probably want to have a `ValidatedProvenance` as well, with a `Name` field of the same `SanitizedName` type. Instances of these types are generated when you load the endorsement and the provenance from file, and you use these objects in the rest of the code.

Reply:
@rbehjati Am I correct in understanding that the goal of this proposal is to ensure that all principal names are sanitized? If that is the goal, isn't it better to just call `SanitizeName` inside the constructor of `Principal`? 

If this is just moved inside the constructor, all the names are still sanitized and the code is changed in just one place to do this. By changing the principal constructor to sanitize, the programmer also never has to think about this whereas if the interface of principal is changed to take a new `SanitizedName` type, the programmer needs to still call `SanitizeName()` everywhere.

Reply:
I guess since `go` doesn't have constructors, the equivalent thing to do would be to make

```
func ConstructPrincipal(contents String) {
   return Principal{ Contents: sanitizeName(contents) }
}
```

and then in the code, call `ConstructPrincipal` everywhere rather than writing `Principal` directly.

Reply:
https://github.com/project-oak/transparent-release/issues/59

Reply:
> isn't it better to just call SanitizeName inside the constructor of Principal?

Then it is really not `sanitizeName` but `sanitizeString`. Is that what you want? Or is `Contents` exactly the same as `Name`? If so, perhaps it should be renamed to `Name`. 

> the programmer needs to still call SanitizeName() everywhere.

No. The programmer only needs to use `SanitizeName()` in the constructor of the types that have an instance of `SanitizeName`. This is a more type-safe approach in my opinion. With your proposal, there will be an inconsistency between the contents of a `Principal` and a related `ValidatedEndorsement`. It might not be an issue now, but it could become an issue as the code evolves. 

I should add that that I don't have a complete model of all the structs and their relations in mind, so maybe you are right and just sanitizing all strings when creating an instance of `Principal` is good enough. 

Please double-check what is the convention for naming constructors in Go. I think I have seen `new`. Also, `Contents` must be changed to `contents` so that it is not exposed. Otherwise, one could just instantiate `Principal` without calling the constructor.

Reply:
If this does not need to be fixed in this PR, and since you both approved the PR, is it OK to move further discussion to that issue and resolve this thread ?

I will speculatively resolve this thread. I added a link to this thread in that issue.

Reply:
Sounds good to me. Thanks :)